src/com/xilinx/rapidwright/rwroute/GlobalSignalRouting.java

```diff
@@ -279,45 +301,141 @@ private static void symmetricClockRoutingVersal(Net clk, Device device, Function
             // we may fail to do so. Thus, we need to force the Y-coordinate of centroid to be 1.
-            assert(source.getTile().getTileYCoordinate() == 0);
-            // And, in X-axis, Vivado doesn't go to the real centroid of target clock regions... it just uses a nearby VROUTE.
-            int centroidX = source.getTile().getClockRegion().getColumn();
-            // VROUTE nodes are in the clock region where X is odd.
-            if (centroidX % 2 == 0) centroidX -= 1;
-            if (centroidX <= 0)     centroidX = 1;
-            centroid = device.getClockRegion(1, centroidX);
-
-            Node clkRoutingLine = VersalClockRouting.routeBUFGToNearestRoutingTrack(clk);// first HROUTE
-            centroidHRouteNode = VersalClockRouting.routeToCentroid(clk, clkRoutingLine, centroid, true);
+            Node clkRoutingLine = VersalClockRouting.routeBUFGToNearestRoutingTrack(clk, getNodeStatus);// first HROUTE
+            Pair<Node, ClockRegion> result = findCentroid(clk, clkRoutingLine, centroid, true,
+                    getNodeStatus, unavailableTracks);
+            centroidHRouteNode = result.getFirst();
+            centroid = result.getSecond();
+        } else if (sourceTypeEnum == SiteTypeEnum.BUFG_PS) {
+            // These source sites are located in the middle of the device. The path from the
+            // output pin to HROUTE matches the following pattern:
+            // NODE_GLOBAL_BUFG (has a suffix "_O_PIN") ->
+            // NODE_GLOBAL_BUFG (the output node with a suffix "_O") ->
+            // NODE_GLOBAL_HROUTE (located in the same clock region)
+            centroidHRouteNode = VersalClockRouting.routeBUFGToNearestRoutingTrack(clk, getNodeStatus);// first HROUTE
         } else {
             throw new RuntimeException("ERROR: Routing clock net with source type " + sourceTypeEnum + " not supported.");
         }
 
-        Node vroute = VersalClockRouting.routeToCentroid(clk, centroidHRouteNode, centroid, false);
+        // If the source and centroid are in the same row, we do not need to traverse vertical routing tracks
+        boolean noVrouteNeeded = centroidHRouteNode.getTile().getClockRegion().getRow() == centroid.getRow();
 
-        Map<ClockRegion, Node> upDownDistLines = VersalClockRouting.routeToHorizontalDistributionLines(clk, vroute, clockRegions, false, getNodeStatus);
+        Pair<Node, ClockRegion> centroidResult = findCentroid(clk, centroidHRouteNode, centroid,
+                noVrouteNeeded, getNodeStatus, unavailableTracks);
+        Node vroute = centroidResult.getFirst();
+        centroid = centroidResult.getSecond();
+        
+        Map<ClockRegion, Node> upDownDistLines = VersalClockRouting
+                .routeToHorizontalDistributionLines(clk, vroute, usedCRsAndNonLCBPinsTuple.getFirst(),
+                        false, getNodeStatus);
+
+        // Route non-LCB driven pins
```

Comment:
Can you give an example of what a non-LCB pin is?

Reply:
I updated the comment to include an example.